### DISCUSSION TITLE
Disable garbage collection temporary

### DISCUSSION THREAD

**Discussion by leiserfg** [L3MON4D3/LuaSnip#818] [Ideas]
Hey @L3MON4D3  I was playing with some optimizations and noticed that disabling the garbage collection while loading vscode snippets (and enabling it after) reduces the time of loading noticeably. The change is more noticeable when using the synchronous `load`.

**Comment by L3MON4D3** [comment]
Interesting! Any downsides to doing that? Do you have some numbers/flamegraph?

**Comment by leiserfg** [comment]
Potential downsides: 
- the user has not enough ram to hold all the stuff at once
- something crashes and we don't re-enable the GC so nvim keeps eating ram forever.

For my measurements, loading goes from btw 0.47 and 0.50  to 0.30

The 0.3 is including the step of re-enabling the GC

**Reply by L3MON4D3** [reply]
Oh, yeah 1 makes sense, ofc
2 is a good point, the most straightforward way to guarantee this seems to be something like
```lua
local no_gc_load = function(...)
    collectgarbage "stop"
    local ok, res = pcall(load, ...)
    collectgarbage "restart"
    if not ok then
        error(res)
    end
end
```

This kind of begs the question if we should just provide a generic interface to this,
where any function can be passed.
But for now, explicit functions without gc seem more comfortable.

**Comment by leiserfg** [comment]
I was trying to profile it, but unfortunately, I could not make it work, the luajit included in nvim for nixos does not include  "jit.profile". Can you do it?
```lua
local prof = require'plenary.profile'

prof.start("/tmp/no_gc_profile.log")
-- collectgarbage "stop"
require("luasnip/loaders/from_vscode").load()
-- collectgarbage "restart"
prof.stop()
```

```lua 
local prof = require'plenary.profile'

prof.start("/tmp/gc_profile.log")
collectgarbage "stop"
require("luasnip/loaders/from_vscode").load()
collectgarbage "restart"
prof.stop()
```

**Reply by L3MON4D3** [reply]
Yes!
with GC:
![a](https://user-images.githubusercontent.com/41961280/224514461-e05bd835-4959-47f3-89e2-f317cfb3b553.svg)
without GC:
![b](https://user-images.githubusercontent.com/41961280/224514468-397a1c5b-d00a-4f01-bc16-9ae9ce305056.svg)

Looks like everything is just faster, nothing in particular catches my eye (I guess that would be expected, more or less?)

Looking at that flamegraph, I think both the deepcopies and `complete_ext_opts` could be optimized quite a bit (though not as easily as just turning of garbage collection for a bit)

**Reply by leiserfg** [reply]
I'm actually doing it for my complete config, it helps not only with luasnip:

```lua
collectgarbage "stop"
-- jit.off()

vim.api.nvim_create_autocmd("InsertEnter", {
  callback = function()
    -- jit.on()
    collectgarbage "restart"
  end,
  once = true,
})

```

as you can see I was also playing with disabling the `jit`, that helps with some plugins but makes it slower for others tho.
